### ISSUE TITLE
Ember: Decoding failure is a server error

### ISSUE THREAD

**Issue by kubukoz** [http4s/http4s#5912]
With the ember server, `CirceEntityCodec`'s default behavior seems to fail any request that doesn't successfully decode, but with an InternalServerError by default.

To reproduce:

`scala-cli https://gist.github.com/kubukoz/351c19621d475c319ab1d0a178bf9bf7`

In another tab:

```http
http post :8080/foo Origin:http://localhost:8080 a=b -v

POST /foo HTTP/1.1
Accept: application/json, */*;q=0.5
Accept-Encoding: gzip, deflate, br
Connection: keep-alive
Content-Length: 10
Content-Type: application/json
Host: localhost:8080
Origin: http://localhost:8080
User-Agent: HTTPie/2.6.0

{
    "a": "b"
}


HTTP/1.1 500 Internal Server Error
Connection: keep-alive
Content-Length: 0
Date: Thu, 20 Jan 2022 16:26:32 GMT

```

I think this is caused by Ember using a default error handler which transforms every throwable into a 500, whereas Blaze's special-cases MessageFailure - is there a specific reason why that's so, or should Ember be updated to behave the same?

**Comment by rossabaker** [comment]
I think Ember should do the same.  Getting correct status codes is why those message failures exist in the first place.

**Comment by ChristopherDavenport** [comment]
I chose not to do that to put more things into user space. Whether or not people want to hold consistent with that is a larger question. The point was to have ember put middlewares in users hands to get normal behavior visible and present in things like metrics, which would otherwise be invisible. 

either the server needs to be more powerful to hold more triggers, or weaker and put more things into user space. I chose to make it do less and put things into user space and added the ErrorHandling middleware so users could do this themselves.

**Comment by rossabaker** [comment]
I agree with the minimalistic server approach, but until those middlewares are discovered and assembled, it's a rough experience for novices.  This is where applying a reasonable set of default middlewares would help.  The idea has been kicked around for years, but is stuck on API: where do they come from?  How do we opt out?  Can we say "give me the defaults, except _that one_"?

This is also where that server TCK would be good, because we advertise that people can switch between backends easily, but there are significant surprises like this.  There's less we can do about that on servlets, but blaze/ember/netty are all low-level enough that we could do better there.

**Comment by ChristopherDavenport** [comment]
I can throw together something for a common Middlewares in a builder pattern approach. Then we can bike shed on the flavor from at least a proposal.

I think if we get this all the backends can simplify, and we'll be consistent.

Not sure what server TCK means sorry.

**Comment by dwalend** [comment]
I'm part-way committed to moving a project from blaze to ember and blundered into this thread while trying to understand a bug.

I have a unit test that shows an http4s service is replying with a 422 when it tries to interpret a "{ }" body. I have an integration test that indicates the ember server is replying with a 500 for the same body (very probably - the system is quite baroque, but the lead hypothesis for the bug is the change from blaze to ember). 

Is this the problem you are discussing here? I'm hopeful but I don't believe it quite matches.
